internal/model/user_op.go

```diff
@@ -36,66 +38,146 @@ type UserOpResult struct {
 
 	// The user was removed from the following courses (by id).
 	Dropped []string `json:"dropped,omitempty"`
+}
+
+// A general representation of the result of operations that modify a user in any way (add, remove, enroll, drop, etc).
+// All user-facing functions (essentially non-db functions) should return an instance or collection of these objects.
+type UserOpResult struct {
+	// Embed BaseUserOpResult for basic fields, see above.
+	BaseUserOpResult
 
 	// The following error occurred during this operation because of the provided data,
 	// i.e., they are caused by the calling user.
 	// All error messages should be safe for users.
-	ValidationErrors []string `json:"validation-errors,omitempty"`
+	ValidationError *LocatableError
 
-	// The following error occurred during this operation, but not because of the provided,
+	// The following error occurred during this operation, but not because of the provided data,
 	// i.e., they are the system's fault.
 	// These errors are not guarenteed to be safe for users,
 	// and the calling code should decide how they should be managed.
-	SystemErrors []string `json:"system-errors,omitempty"`
+	SystemError *LocatableError
+
+	// The following error occurred during this operation, but not because of the provided data,
+	// i.e., the system was unable to communicate the results.
+	// These errors are not guarenteed to be safe for users,
+	// and the calling code should decide how they should be managed.
+	CommunicationError *LocatableError
 
 	// The following cleartext password was generated during this operation.
 	// Care should be taken to not expose their field.
```

Comment:
s/their/this/

I know this is not your text, but go ahead and change it while you are in here.
Thanks.